### PR TITLE
feat(generation): carry the wall pattern through compartment dividers

### LINKED ISSUE #2811: Bin Divider Wall and Floor Patterns
**Problem**
The designer currently only generates solid bin divider walls. This is sometimes not needed or not desired, for example to improve visibility or ventilation for the compartments. It also doesn't look particularly balanced having a hollowed out outer wall with all solid dividers.

**Proposed Solution**
Allow the user to choose to enable patterned bin divider walls, using the same wall pattern logic as that of the existing bin wall chooser under Shape>Wall Pattern and Shape>Wall Pattern Scale. Perhaps there could be a sync button that forces the same pattern among the outer and divider walls.

**Alternatives Considered**
Additionally it could be helpful to also allow patterns in the base. I have had to do this manually for a few washroom drainer bins, and I imagine there must be other users who could use patterned floors.

### PR DESCRIPTION
**Pull request by andymai**
Closes #2811 for integrated compartment dividers.

## Problem

The designer only ever generated solid divider walls. A patterned bin therefore read as hollowed-out outer walls wrapped around solid dividers — visually unbalanced, and no help for the visibility or ventilation the reporter wanted inside each compartment.

## What this does

Adds a **Pattern divider walls** checkbox under Shape → Wall Pattern. When on, the compartment dividers carry the same pattern *and* the same scale as the outer walls. There is deliberately no second picker and no sync button: the two can never be out of sync because there is only one setting.

Holes cut fully through, matching the outer walls — that is what delivers the see-through visibility and airflow. Users who need compartments sealed simply leave it off.

Both pattern pipelines are supported, including kumiko.

## How it works

- `dividerPatterns.ts` (pure) resolves which dividers can be patterned, the band each offers, and the keep-out rectangles where it must stay solid. Keep-outs are found by projecting each intruding feature's world footprint onto the divider centerline, so tilted dividers need no special case.
- `dividerPatternBuilder.ts` builds every panel in the divider's **local frame** and places it with one rotate + translate. Two dividers with the same span and keep-outs produce identical geometry and share a cache entry, wherever they sit and whatever their tilt.
- Stamp patterns honour keep-outs by dropping whole **elements** — no boolean, and no half-hex knife edge that would print as an unsupported fin.
- Kumiko panels resolve their lattice against the **outer perimeter** (`resolveKumikoPerimeter`) and take a u-window of it. That is legal because the lattice is u-periodic, and it means a divider's triangles come out exactly the size the walls carry. Only the band height is re-fitted.

## Decisions worth flagging

- **Shortened dividers re-fit the band** rather than clipping the wall band at the divider top. Clipping would slice the top element row, leaving the unsupported webs that `BOTTOM_SOLID_SKIRT` exists to prevent (#2317).
- **Cut depth is `thickness + 2`, not the outer walls' `wallThickness * 4`.** The narrowest legal compartment is only `2 x thickness` wide, so a deep prism would reach across and perforate the *next* divider at meaningless positions.
- **Keep-outs cover every feature that bonds to a divider**: scoop ramps, label tab shelves and gussets, interior cutouts, divider crossings, and wall junctions. Label tabs use the compartment's full width rather than the resolved tab width — conservative in the safe direction, and it avoids duplicating the tab-width capping math.
- **Print estimates and the generation watchdog were updated too.** Without a per-divider term a dense multi-compartment bin can exhaust the flat pattern budget mid-generation and hard-reset the worker; kumiko divider panels are charged much more than stamp panels.
- The UI's too-small note comes from `utils/dividerPatternFit.ts`, a deliberately conservative params-only mirror that judges the band and each divider's span but not the keep-outs. A divider cleared entirely by keep-outs still stays solid silently — the same convention every other pattern feature follows.

## Not in this PR

Slotted bins' **removable divider pieces** are a separate geometry path (`dividerBuilder` + `dividerExport`, with cross-lap notches and tab engagement to keep clear of) and follow in a second PR. The checkbox is disabled with an explanation on those bins.

Floor / base patterns from the issue's "Alternatives Considered" are tracked separately — a drainage hole must pass through the base socket, which collides with the deferred-socket preview optimisation and deserves its own design.

## Testing

- New `binGenerator.scenario.dividerPatterns` suite (15 cases, real OCCT): volume comparisons prove material is actually removed and that a runaway prism is not; plus scoops, interior cutouts, label tabs, outer cutouts, tilted dividers, shortened dividers, half-grid, overhang, and mitsukude lattice panels. Two inertness cases pin that a bin with no dividers — and a divider with no room for a band — come out byte-identical with the option on.
- 16 unit tests on the placement planner, 9 on the UI predicate, plus migration, timeout and print-estimate coverage.
- Adjacent scenario suites (wallPatterns, kumiko, honeycombJunction, compartments) re-run with no snapshot churn; 3609 bin-designer / bridge unit tests pass.
- `typecheck`, `lint`, `knip`, `check:boundaries`, `check:i18n`, `check:i18n:values` all clean.

## Screenshots

Not included — this is geometry that needs a real 3D preview to judge. Happy to attach renders of a 2x2 honeycomb bin and a mitsukude bin before merge if you'd like to eyeball the divider panels.

---
## Summary by cubic
Adds an option to carry the wall pattern through compartment dividers so patterned bins are see‑through and balanced. Dividers reuse the outer walls’ pattern and scale; supports both stamp and kumiko, addressing #2811.

- **New Features**
  - UI: Added “Pattern divider walls” under Shape → Wall Pattern with availability/too‑small hints; disabled on custom shapes and removable dividers.
  - Geometry: Dividers use the same pattern and scale; stamp drops whole elements near keep‑outs; kumiko panels resolve against the outer perimeter for matching triangle size; band re‑fitted for shortened dividers; tilted dividers supported; cut depth set to thickness + 2; keep‑outs cover scoops, label tabs, interior cutouts, crossings, and wall junctions.
  - Pipeline/tests: New divider pattern planner + builder wired into the features stage; scenarios and unit tests added; i18n/docs updated; `wallPattern.dividers` param with migration backfill to false; shared `wallPatternMetrics` facade for element sizing; print estimates include divider perforations; generation watchdog adds a per‑divider budget, gated to standard, non‑solid, non‑polygon bins and by divider thickness.

- **Bug Fixes**
  - Estimates: Use the real divider height (accounts for interior taper and `compartments.dividerHeight`), and apply the full worker gate for divider pattern reduction (polygon footprint, solid base, zero thickness) so estimates don’t claim savings when no panels are generated.
  - Gating/timeouts: Divider thickness gating mirrored across UI fit check, checkbox availability, timeout budgeting, and estimates; divider timeout bonus applies only when eligible; kumiko timeout test now exercises a kumiko config.
  - Robustness/memory: Keep‑out projection uses an epsilon for near‑parallel cases to avoid numerical edge conditions; divider panel fuses are kept inside their disposal guard to prevent handle leaks on `fuseAllOrNull` throws, with disposal folded into a single helper.
  - UI copy: Solid bins now show a dedicated explanation for why divider patterning is unavailable (instead of the slotted‑dividers message).

<sup>Written for commit 9842447d4a692b7d6ba09a42d0c162a7b39a49ab. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/andymai/gridfinity-layout-tool/pull/2814?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/features/bin-designer/utils/printEstimates.ts [hunk 1]
```diff
@@ -182,15 +183,22 @@ function computeDividerVolume(
   const innerW = outerW - 2 * wallThickness;
   const innerD = outerD - 2 * wallThickness;
   const dividerH = totalH - bottomH;
-  const { cols, rows, thickness, cells } = params.compartments;
+  const { cols, rows, thickness } = params.compartments;
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/bin-designer/utils/printEstimates.ts:205]: Good catch — fixed in 92a3b27.

You are right on both counts, and the consistency half is the one this PR created: `dividerPatternReduction` resolves the effective divider height while `computeDividerVolume` did not, so the two terms were scaled against different bases. The pre-existing half is worse in practice — shortening the dividers never lowered the estimate at all, because the height came from `totalH - BASE_HEIGHT` regardless.

Both now share a single `effectiveDividerHeight()` helper that mirrors the generator: cavity floor to interior ceiling (lip taper included via `computeInteriorHeight`), capped by `compartments.dividerHeight`. Added a test pinning that a shortened divider actually reduces the estimate.

— claude

File: src/features/bin-designer/utils/dividerPatternFit.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+/** Assess how much of the divider pattern this bin can actually carry. */
+export function assessDividerPatternFit(params: BinParams): DividerPatternFit {
+  const { wallPattern } = params;
+  if (!wallPattern.enabled || wallPattern.dividers !== true) return 'unavailable';
+  if (params.style !== 'standard') return 'unavailable';
+  if (params.base.solid) return 'unavailable';
+  if (isPartialMask(params.cellMask)) return 'unavailable';
+
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/bin-designer/utils/dividerPatternFit.ts:89]: Fixed in 0142f3c — `assessDividerPatternFit` now returns `unavailable` for `compartments.thickness <= 0`, matching `dividerPatternsApply`.

Worth noting it is reachable: `migrateParams` spreads a persisted `compartments` object without clamping thickness, so a crafted or legacy payload can carry `0` even though the UI never produces it. Test added.

— claude

File: src/features/bin-designer/components/panel/WallsSection/useWallsSection.ts [hunk 3]
```diff
@@ -87,6 +89,30 @@ export function useWallsSection() {
...
+    [updateWallPattern]
+  );
+
+  const dividersAvailableReason = useMemo(() => {
+    if (params.style !== 'standard') return t('binDesigner.walls.pattern.dividers.notStandard');
+    if (isPartialMask(params.cellMask)) return t('binDesigner.walls.pattern.dividers.notPolygon');
+    if (getCompartmentCount(params.compartments) <= 1)
+      return t('binDesigner.walls.pattern.dividers.noDividers');
+    return undefined;
+  }, [params.style, params.cellMask, params.compartments, t]);
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/bin-designer/components/panel/WallsSection/useWallsSection.ts]: Fixed in 0142f3c — the availability check is now `getCompartmentCount(...) <= 1 || params.compartments.thickness <= 0`, so a zero-thickness grid reads as "no dividers to pattern" rather than offering an option the worker would discard. Test added.

— claude

File: src/features/generation/bridge/generationTimeout.ts [hunk 4]
```diff
@@ -183,6 +220,15 @@ function binRawBudgetMs(params: BinParams): number {
...
     }
+
+    if (params.wallPattern.dividers === true) {
+      const segments = countDividerSegments(params);
+      timeout +=
+        segments *
+        (isKumikoPattern(params.wallPattern.pattern)
+          ? KUMIKO_DIVIDER_MS_PER_SEGMENT
+          : DIVIDER_PATTERN_MS_PER_SEGMENT);
+    }
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/generation/bridge/generationTimeout.ts:241]: Fixed in 0142f3c — `countDividerSegments` now mirrors `dividerPatternsApply` up front: it returns 0 for slotted and solid bins, polygon (cellMask) footprints, and zero-thickness grids, so the budget only pays for panels the worker will actually build. Test asserts each of those cases lands exactly on the no-divider baseline.

— claude

File: src/features/generation/worker/generators/dividerPatterns.ts [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+    [-dx, x0 - box.xMin],
+    [dx, box.xMax - x0],
+    [-dy, y0 - box.yMin],
+    [dy, box.yMax - y0],
+  ];
+  for (const [p, q] of edges) {
+    if (p === 0) {
+      if (q < 0) return null;
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/generation/worker/generators/dividerPatterns.ts:132]: Changed to an epsilon in 8e4b58a — you are right that the branch was unreachable: `p` derives from `Math.cos`, and `cos(90°)` is 6.1e-17, so a column divider never took the parallel path.

One correction to the framing though, because I tried to write a regression test for it and could not: I probed the plan output with and without the epsilon on a column divider whose scoop box edge sits exactly on it (the degenerate input), and the result is **byte-identical**. The reason is that the two candidate boxes either side of a divider are symmetric in `u`, so whichever one the dust-sign happens to admit yields the same range.

So this is a latent-robustness fix, not an observable defect — dividing dust by dust to decide a clip is a hazard waiting for a future asymmetric keep-out, but nothing today reaches it. I have not attached a test that claims otherwise; the case that *is* pinned is that a coincident box still anchors to the divider end.

— claude

File: src/features/generation/bridge/generationTimeout.test.ts [hunk 6]
```diff
@@ -240,6 +242,59 @@ describe('computeGenerationTimeoutMs', () => {
...
+    const compartments = { cols: 2, rows: 1, cells: [0, 1], thickness: 1.2 };
+    const stamp = computeGenerationTimeoutMs(
+      params({ wallPattern: { ...HEX_ON, dividers: true }, compartments })
+    );
+    const stampOff = computeGenerationTimeoutMs(
+      params({ wallPattern: { ...HEX_ON, dividers: false }, compartments })
+    );
+    expect(stamp - stampOff).toBe(DIVIDER_PATTERN_MS_PER_SEGMENT);
+    expect(KUMIKO_DIVIDER_MS_PER_SEGMENT).toBeGreaterThan(DIVIDER_PATTERN_MS_PER_SEGMENT);
+  });
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/generation/bridge/generationTimeout.test.ts]: Fair — that was asserting a relationship between two constants, not the branch. Fixed in 8e4b58a: the test now runs both a stamp and a kumiko config through `computeGenerationTimeoutMs` on a single-divider grid, so each delta is exactly one segment charge and pins that `isKumikoPattern(...)` selects the right constant.

— claude

File: src/features/bin-designer/utils/printEstimates.ts [hunk 7]
```diff
@@ -548,7 +571,39 @@ function computeWallPatternReduction(
...
+ */
+function dividerPatternReduction(
+  params: BinParams,
+  innerW: number,
+  innerD: number,
+  coverageFraction: number
+): number {
+  if (params.wallPattern.dividers !== true) return 0;
+  if (params.style !== 'standard') return 0;
+
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/bin-designer/utils/printEstimates.ts:601]: Fixed in 8e4b58a — `dividerPatternReduction` now applies the whole `dividerPatternsApply` gate: polygon footprint, solid base, and zero divider thickness, on top of the existing style check.

The polygon case is the one that actually bites, since a design can carry `dividers: true` from before a shape change and the checkbox being disabled does not clear the stored flag. Test added asserting the estimate is unchanged for a cellMask bin.

— claude

File: src/features/generation/worker/generators/dividerPatternBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+  bandCenterZ: number,
+  depth: number
+): Shape3D | null {
+  const boxes = buildKeepOutBoxes(keepOuts, bandCenterZ, depth);
+  if (boxes.length === 0) return panel;
+  const tool = fuseAllOrNull(boxes);
+  if (!tool) {
+    for (const b of boxes) b.delete();
+    return panel;
+  }
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/generation/worker/generators/dividerPatternBuilder.ts:163]: Correct on both counts — fixed in 39e819c.

Slightly embarrassing one: the comment two lines below the second site says leaking WASM handles on the error path is what wedges the worker heap, and the fuse was sitting right outside the guard doing exactly that. Both call sites now have the `fuseAllOrNull` inside the try, so the boxes / window pieces are disposed whether it returns a shape, returns null, or throws.

Also folded the duplicated inline `try { s.delete() } catch {}` blocks into a local `disposeQuiet`, matching `wallPatternClips`.

— claude

File: src/features/bin-designer/components/panel/WallsSection/useWallsSection.ts [hunk 9]
```diff
@@ -87,6 +89,32 @@ export function useWallsSection() {
...
+
+  const dividersAvailableReason = useMemo(() => {
+    if (params.style !== 'standard') return t('binDesigner.walls.pattern.dividers.notStandard');
+    if (isPartialMask(params.cellMask)) return t('binDesigner.walls.pattern.dividers.notPolygon');
+    // Zero thickness means compartment IDs with no wall between them — nothing
+    // to pattern, and the worker gate rejects it too.
+    if (getCompartmentCount(params.compartments) <= 1 || params.compartments.thickness <= 0)
+      return t('binDesigner.walls.pattern.dividers.noDividers');
+    return undefined;
+  }, [params.style, params.cellMask, params.compartments, t]);
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/bin-designer/components/panel/WallsSection/useWallsSection.ts]: Partly — the conclusion is right but the mechanism is not, so worth writing down.

The checkbox is **not** enabled on solid bins today. `IMPLICATION_RULES` in `shared/constraints/rules.ts` keeps `base.solid` and `style === "solid"` in lockstep in both directions, so a solid bin always has `style === "solid"` and the existing `style !== "standard"` check already disables it.

What you did surface is real though: it falls through to the **wrong copy**. A solid bin was being told "Removable dividers print as separate pieces and stay solid", which describes slotted bins — a solid bin has no interior at all. So the user got an explanation, just a false one.

Fixed in 9842447 with a dedicated `dividers.notSolid` string ("Solid bins have no compartments to divide") checked ahead of the slotted branch, and `base.solid` added to the condition so the hook mirrors `assessDividerPatternFit`, which already gated on it. Translated across all 10 locales; test added.

— claude
